peekingduck/pipeline/nodes/model/jde_mot/jde_files/matching.py

```diff
@@ -0,0 +1,189 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Original copyright (c) 2019 ZhongdaoWang
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+Functions for bounding box matching
+"""
+
+from typing import Any, List, Tuple, Union
+import lap
+import numpy as np
+from scipy.spatial.distance import cdist
+from cython_bbox import bbox_overlaps as bbox_ious
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files.utils import kalman_filter
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files.utils.kalman_filter import (
+    KalmanFilter,
+)
+
+# pylint: disable=invalid-name, no-name-in-module, unused-argument, redefined-outer-name, too-many-arguments
+
+
+def linear_assignment(
+    cost_matrix: np.ndarray, thresh: float
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Function for Hungarian Algorithm association.
+
+    Args:
+        cost_matrix (np.ndarray): Cost matrix of current and previous matrices.
+        thresh (float): Threshold value
+
+    Returns:
+        (Tuple[np.ndarray, np.ndarray, np.ndarray]): Returned tuple
+            contains arrays of matched and unmatched tracks.
+    """
+    if cost_matrix.size == 0:
+        return (
+            np.empty((0, 2), dtype=int),
+            tuple(range(cost_matrix.shape[0])),
+            tuple(range(cost_matrix.shape[1])),
+        )
+    matches, unmatched_a, unmatched_b = [], [], []
+    _, x, y = lap.lapjv(cost_matrix, extend_cost=True, cost_limit=thresh)
+    for ix, mx in enumerate(x):
+        if mx >= 0:
+            matches.append([ix, mx])
+    unmatched_a = np.where(x < 0)[0]
+    unmatched_b = np.where(y < 0)[0]
+    matches = np.asarray(matches)
+
+    return matches, unmatched_a, unmatched_b
+
+
+def ious(
+    atlbrs: Union[List[Any], np.ndarray], btlbrs: Union[List[Any], np.ndarray]
```

Comment:
`atlbrs` is of type `List[STrack]`. However `STrack` cannot be imported into the module due to an `ImportError (most likely due to a circular import)` thus Any is used in case.

Reply:
For function docstring, would it be best to have a type hint of `STrack` or `Any` ?

Reply:
After discussion, Any to be used in variable type hint as well as in function docstring. STrack will be used in description.